### DISCUSSION TITLE
external_connectivity

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/labview_book#25] [Announcements]
# external_connectivity

连接其它外部程序或组件

https://lv.qizhen.xyz/external_connectivity

**Comment by AbetterLee** [comment]
请问事件回调注册的event in是怎么创建的，里面的post event.vi在哪里找的？

**Reply by ruanqizhen** [reply]
事件回调注册函数的“用户参数”可以是用户定义的任何类型的数据，使用者可以根据自己的需要放任何数据进去。在这里使用 Event 类数据是为了把ActiveX 控件的事件，被转换成了 LabVIEW 的用户自定义事件，使得最终代码更易读。
Event 类数据的创建和使用可以参考：https://lv.qizhen.xyz/pattern_ui#%E9%80%9A%E7%94%A8%E7%94%A8%E6%88%B7%E8%87%AA%E5%AE%9A%E4%B9%89%E4%BA%8B%E4%BB%B6%E7%9A%84%E8%AE%BE%E8%AE%A1

**Reply by AbetterLee** [reply]
这个理解了，谢谢老师。还有一个地方不明白，创建的回调vi在程序框图里没有体现，实际实验时代码里还是用的事件回调注册，请问是这样用吗？

**Reply by ruanqizhen** [reply]
回调vi不需要被显式的调用，VI 被注册之后，有ActiveX 事件出现，回调VI就会自动被调用。 另外，如果这里的截图不够详细，也可以直接打开示例VI查看一下。

**Comment by c572292497** [comment]
大佬，我想问一下，ActiveX控件的线程怎么设置呢？

**Reply by ruanqizhen** [reply]
ActiveX控件的线程不需要什么特殊设置。作为放置在前面板上的控件，只要使用默认的界面线程来调用ActiveX控件就可以了。有些ActiveX控件可能有自己内部的线程管理，也不需要要在LabVIEW中设置。

**Reply by c572292497** [reply]
我了解了，感谢大佬。然后我现在还有个问题，我希望通过循环事件结构来触发 开始采集实验数据的模块。但是采集实验数据需要循环采集，在循环的过程中我不能点击前面板上的停止按钮来结束采集，请问有什么别的比较好的方法解决嘛？

**Reply by ruanqizhen** [reply]
采集数据的部分需要运行在另一个线程，这样才不会阻塞界面线程的事件。关于把子任务加载到新线程，可以参考一下 [https://lv.qizhen.xyz/vi_server_for_subvi](https://lv.qizhen.xyz/vi_server_for_subvi) 中的“动态调用 VI”和“异步调用”。

**Comment by Arita9527** [comment]
老师，为啥有时候ACTIVE需要Variant To Data Function 控件来指定类型？如果不知道类型怎么办？

**Reply by ruanqizhen** [reply]
函数（和 VI）之间的数据传递需要至少两个信息，数据所在的内存地址和数据的类型。同一个语言内部，数据类型也许不需要额外说明，因为在做类型定义，或者使用数据控件等的时候，这些信息就已经在里面了。但是跨平台调用的时候，另一放没有这些信息，就需要额外把数据类型的信息也告诉对方。

**Reply by Arita9527** [reply]
大概明白了，
但是我怎么知道它传出来是什么数据类型呢？到底是连接字符串，还是引用？
有时候引用还要右键选择一个对象？

**Reply by ruanqizhen** [reply]
这个必须要查看文档了。如果一个ActiveX控件是提供给用户使用的，肯定附带文档说明，解释每个函数的功能、参数和数据类型等。

**Comment by ruanqizhen** [comment]
之前博客上的相关文章，搬到这里：

VB Script 打开一个VI
2008年1月9日 · 阅读需 1 分钟
    LabVIEW 的一些服务功能是通过 ActiveX 接口提供出来的，我们在其他支持 ActiveX 的语言中可以方便的调用这些服务。比如一个最简单的例子，使用 VB Script 打开一个 VI 的前面板，这样做就行了：

Set lvapp = CreateObject("LabVIEW.Application")
Set vi = lvapp.GetVIReference("C:\temp\test.vi")
vi.FPWinOpen = True

    由于 IE 支持 VB Script，这段代码还可以嵌在 HTML 文件中实现这样的功能：页面上有一处超级链指向一个VI，点击这个链接，就可以显示相应的 VI。其他浏览器可以使用 JavaScript。

**Reply by ruanqizhen** [reply]
之前博客下的留言：

Qizhen
2008年01月19日 12:48 编辑
你的 LabVIEW 是什么版本的？我在 8.5 上试验了一下，如果设置为可重入，是会有两份界面打开。要注意的是，这个界面要被设计成一个子VI，不能是最顶层的VI。如果你是用的是 LabVIEW 8.0 以上的版本，还可以使用XControl。把所有界面部分做成一个 XControl，然后连个VI共享这一份 XControl。

回复
xm
2008年01月14日 18:44 编辑
谢谢你的回复
我一开始也是看了你blog上写的有关可重入部分内容后，想着把子VI设为可重入，然后用Open VI Reference去两次调用的，但是好像还是只能有一个界面。LabVIEW跟其他应用程序好像有点不一样，用LabVIEW生成的exe文件，在一台机器上只能运行一个进程，当已经运行一个exe文件后，再去打开它，返回的还是先前的运行进程。同样，调用一个子VI时，似乎LabVIEW会先检查内存中是否存在这个子VI，如果存在，还是返回这个子VI的运行界面的（即便设为可重入，当然可能内存中会有两份存在，但是界面我试了一下，确实还是一个相同的界面的）。而我想实现的功能是：有两个按钮，点击第一个按钮后，运行子VI，弹出这个子VI相应的界面（当然它是非模态窗口），然后再点击第二个按钮，同样调用这个子VI，但是要弹出另外一个相同的界面。当然，用不同的子VI名是可以实现这个功能的，但是考虑代码维护方便，同时让相同的代码用不同的名称也有点让人别扭，所以请教博主是否有更好的方法。谢谢

回复
Qizhen
2008年01月14日 15:00 编辑
回xm。如果你就是需要两份同样的界面，那最简单的方法，做两个界面一样的VI。如果是需要打开不定份数的相同界面，把那个子VI设为可重入即可。

回复
xm
2008年01月12日 18:49 编辑
或者把子VI做成exe，我在LabVIEW中同时调用这个exe文件，但是要能同时显示两个界面。

回复
xm
2008年01月12日 18:47 编辑
请问:LabVIEW能否在一个程序中同时调用一个同名的子VI,同时显示这个子VI的两个相同界面。请博主有空解答一下，谢谢。

回复
永青
2008年01月10日 09:58 编辑
这个有点意思，我要试试

回复
WL
2008年01月9日 22:51 编辑
…难….

回复
